### DISCUSSION TITLE
Has anyone had any luck with mono / pythonnet ?

### DISCUSSION THREAD

**Discussion by cellularmitosis** [Tyrrrz/YoutubeExplode#691] [Q&A] [answered]
I'm trying to get [pythonnet](http://pythonnet.github.io/) to load YoutubeExplode via [mono](https://www.mono-project.com/) on Ubuntu Linux.

After `nuget install YoutubeExplode`, I'm trying the following:

```python
#!/usr/bin/env python3

# See https://pythonnet.github.io/pythonnet/python.html
# See https://github.com/Tyrrrz/YoutubeExplode

import os

import pythonnet
pythonnet.load("mono")

import clr

import System
for s in [
    "AngleSharp",
    "Microsoft.Bcl.AsyncInterfaces",
    "System.Buffers",
    "System.Memory",
    "System.Numerics.Vectors",
    "System.Runtime.CompilerServices.Unsafe",
    "System.Text.Encoding.CodePages",
    "System.Text.Encodings.Web",
    "System.Text.Json",
    "System.Threading.Tasks.Extensions",
    "System.ValueTuple",
    "System.ValueTuple",
    "YoutubeExplode",
]:
    clr.AddReference(s)

import YoutubeExplode

youtube = YoutubeExplode.YoutubeClient()
print(youtube)
```

which results in this error about `advapi32.dll`:

```
System.DllNotFoundException: advapi32.dll assembly:<unknown assembly> type:<unknown type> member:(null)
  at (wrapper managed-to-native) FuckRussia.<Initializer_Utils>FD0D4E2FDD637F9F8652DA673240B9A05FD550E7BDEAE27A664F37AD325DBAE02__NativeMethods.RegOpenKeyEx(uintptr,string,int,int,uintptr&)
  at FuckRussia.Initializer.GetCurrentUserRegistryValue (System.String key, System.String entry) [0x0000a] in <22389cef2758435dab31accd307884f1>:0 
  at FuckRussia.Initializer.IsRestricted () [0x00031] in <22389cef2758435dab31accd307884f1>:0 
  at FuckRussia.Initializer.Execute () [0x00007] in <22389cef2758435dab31accd307884f1>:0 
  at <Module>..cctor () [0x00000] in <22389cef2758435dab31accd307884f1>:0 
```

It appears this is a system-level Windows dll, so I'm not sure if mono is able to provide an equivalent.

Anyway, thought I'd start a thread and see if anyone else has had luck with this, or with running on mono in general.

Cheers!

**Comment by cellularmitosis** [comment]
I should mention that I have no idea what I'm doing with either mono or C#, so if you see a seemingly obvious error, please point it out!

**Reply by Sekoree** [reply]
Is there any reason you need to call it through mono? Why not just use regular dotnet and compile for Linux?

**Reply by cellularmitosis** [reply]
I didn't realize mono wasn't the only option for Linux, ha!

**Comment by Tyrrrz** [comment, marked as answer]
The error comes from here: https://github.com/Tyrrrz/FuckRussia/blob/27043a5be63fa54cccc15c0dee2bf4fc54a768b5/FuckRussia/Initializer.cs#L49-L57

It appears mono might be lying what OS it's running on, possible to ensure compatibility with legacy .NET Framework.

That said, I don't see why you'd need to use mono at all, you should be able to run your code with .NET directly: https://dotnet.microsoft.com/en-us/download

I haven't used Python.NET but I assume you should replace your `load(...)` call with the following:

```diff
import pythonnet
-pythonnet.load("mono")
+pythonnet.load("coreclr")
```

Make sure you have the latest .NET SDK installed.

**Reply by cellularmitosis** [reply]
Ah thanks!  Bit embarrassed to say my knowledge of C# is so old that I still assumed mono was the only option for Linux.

After [installing](https://learn.microsoft.com/en-us/dotnet/core/install/linux-ubuntu) the microsoft packages,
I had to perform a bit of a [fixup](https://stackoverflow.com/a/73899341), then I was able to run this:

```python
#!/usr/bin/env python3

# See https://pythonnet.github.io/pythonnet/python.html
# See https://github.com/Tyrrrz/YoutubeExplode

# bash:
# nuget install YoutubeExplode
# ln -s AngleSharp.1.0.1/lib/net7.0/*.dll .
# ln -s YoutubeExplode.6.2.8/lib/net5.0/*.dll .

import os

import pythonnet
pythonnet.load("coreclr")
import clr

clr.AddReference("AngleSharp")
clr.AddReference("YoutubeExplode")
import YoutubeExplode

v_id = YoutubeExplode.Videos.VideoId("dQw4w9WgXcQ")
client = YoutubeExplode.YoutubeClient()
v = client.Videos.GetAsync(v_id);
v.Finalize()
print(v.Result)
```

and I got a result!

```
$ ./yte.py 
Video (Rick Astley - Never Gonna Give You Up (Official Music Video))
```

Thanks so much for the pointer!

Not quite sure how to deal with async function invocation across pythonnet, maybe there is a more idiomatic way than using Finalize and Result.

**Reply by cellularmitosis** [reply]
ah, you can just tack `.Result` onto the end of the call.

```python
v = client.Videos.GetAsync(v_id).Result
print(v.Title)
```

**Reply by Tyrrrz** [reply]
I think Python has `async`/`await`: https://docs.python.org/3/library/asyncio-task.html

Maybe Python.NET bridges those concepts
